feat: Components rewrite (#1336)
discordjs/guide: guide/message-components/interactions.md

# Component interactions

Every button click or select menu selection on a component sent by your bot fires an `interaction`, triggering the <DocsLink path="class/Client?scrollTo=e-interactionCreate" /> event. How you decide to handle this will likely depend on the purpose of the components. Options include:

> - Waiting for a single interaction via <DocsLink path="class/InteractionResponse?scrollTo=awaitMessageComponent" />.

Reviewer: ```suggestion
- Waiting for a single interaction via <DocsLink path="class/InteractionResponse?scrollTo=awaitMessageComponent" type="method"/>.
```

Once `update()` has been called, future messages can be sent by calling `followUp()` or edits can be made by calling `editReply()` on the component interaction.

### Deferred updates

Responding to a component interaction via the `deferUpdate()` method acknowledges the interaction and resets the message state. This method can be used to suppress the need for further responses, however it's encouraged to provide meaningful feedback to users via an `update()` or ephemeral `reply()` at least.

Once `deferUpdate()` has been called, future messages can be sent by calling `followUp()` or edits can be made by calling `editReply()` on the component interaction.

## Awaiting components

If you followed our [buttons](/message-components/buttons) guide, the confirmation workflow for the `ban` command is a good example of a situation where your bot is expecting to receive a single response, from either the Confirm or Cancel button.

> Begin by storing the <DocsLink path="class/InteractionResponse" /> as a variable, and calling <DocsLink path="class/InteractionResponse?scrollTo=awaitMessageComponent" /> on this instance. This method returns a Promise that resolves when any interaction passes it's filter (if one is provided), or throws if none are received before the timeout. If this happens, remove the components and notify the user.

Reviewer: ```suggestion
Begin by storing the <DocsLink path="class/InteractionResponse" /> as a variable, and calling <DocsLink path="class/InteractionResponse?scrollTo=awaitMessageComponent" type="method" /> on this instance. This method returns a Promise that resolves when any interaction passes it's filter (if one is provided), or throws if none are received before the timeout. If this happens, remove the components and notify the user.
```

> Begin by storing the <DocsLink path="class/InteractionResponse" /> as a variable, and calling <DocsLink path="class/InteractionResponse?scrollTo=awaitMessageComponent" type="method" /> on this instance. This method returns a Promise that resolves when any interaction passes it's filter (if one is provided), or throws if none are received before the timeout. If this happens, remove the components and notify the user.

Reviewer: ```suggestion
Begin by storing the <DocsLink path="class/InteractionResponse" /> as a variable, and calling <DocsLink path="class/InteractionResponse?scrollTo=awaitMessageComponent" type="method" /> on this instance. This method returns a [Promise](/additional-info/async-await.md) that resolves when any interaction passes its filter (if one is provided), or throws if none are received before the timeout. If this happens, remove the components and notify the user.
```

```js {10-15}
const response = await interaction.reply({
	content: `Are you sure you want to ban ${target.username} for reason: ${reason}?`,
	components: [row],
});

const filter = i => i.user.id === interaction.user.id;
try {
	const confirmation = await response.awaitMessageComponent({ filter, time: 60_000 });

	if (confirmation.customId === 'confirm') {
		await interaction.guild.members.ban(target);
		await i.update({ content: `${target.username} has been banned for reason: ${reason}`, components: [] });
	} else if (confirmation.customId === 'cancel') {
		await i.update({ content: 'Action cancelled', components: [] });
	}
} catch (e) {
	await response.editReply({ content: 'Confirmation not received within 1 minute, cancelling', components: [] });
}
```

## Component collectors

For situations where you want to collect multiple interactions, the Collector approach is better suited than awaiting singular interactions. Following on from the [select menus](/message-components/select-menus) guide, you're going to extend that example to use an <DocsLink path="class/InteractionCollector"/> to listen for multiple <DocsLink path="class/StringSelectMenuInteraction"/>s.

> Begin by storing the <DocsLink path="class/InteractionResponse" /> as a variable, and calling <DocsLink path="class/InteractionResponse?scrollTo=createMessageComponentCollector" /> on this instance. This method returns an InteractionCollector that will fire it's <DocsLink path="class/InteractionCollector?scrollTo=e-collect" /> event whenever an interaction passes its filter (if one is provided).

Reviewer: ```suggestion
Begin by storing the <DocsLink path="class/InteractionResponse" /> as a variable, and calling <DocsLink path="class/InteractionResponse?scrollTo=createMessageComponentCollector" type="method" /> on this instance. This method returns an InteractionCollector that will fire it's <DocsLink path="class/InteractionCollector?scrollTo=e-collect" /> event whenever an interaction passes its filter (if one is provided).
```

> Begin by storing the <DocsLink path="class/InteractionResponse" /> as a variable, and calling <DocsLink path="class/InteractionResponse?scrollTo=createMessageComponentCollector" type="method" /> on this instance. This method returns an InteractionCollector that will fire it's <DocsLink path="class/InteractionCollector?scrollTo=e-collect" /> event whenever an interaction passes its filter (if one is provided).

Reviewer: ```suggestion
Begin by storing the <DocsLink path="class/InteractionResponse" /> as a variable, and calling <DocsLink path="class/InteractionResponse?scrollTo=createMessageComponentCollector" type="method" /> on this instance. This method returns an InteractionCollector that will fire its <DocsLink path="class/InteractionCollector?scrollTo=e-collect" /> event whenever an interaction passes its filter (if one is provided).
```
